src/rt-app.c

```diff
@@ -976,6 +976,23 @@ static void set_thread_priority(thread_data_t *data, sched_data_t *sched_data)
 		sa_params.sched_flags = SCHED_FLAG_KEEP_ALL;
 		tid = gettid();
 
+		/*
+		 * Currently we cannot use the syscall to set a clamp
+		 * value without specifing a valid policy and priotiry.
+		 * Read the current values and use them for the following
+		 * syscall.
+		 * TODO: fix __sched_setscheduler() to get rid of this!
+		 */
+		ret = sched_getattr(tid, &sa_params,
```

Comment:
Can't you get the data directly from data->curr_sched_data ?
When you set only uclamp_{min,max}, we create a sched_data_t with policy == same == -1
this will be replaced by current sched_policy

Also I'm not sure of the behavior at runtime.
It is assumed that if policy == same then we have a priority set correctly in the sched_data which is not the case if only uclamp_min/max has been set

Reply:
When you say policy == same == -1; do you mean from rt-app point of view or the kernel?

AFAIU the problem is that to do the sched_setattr() syscall, the policy can't be -1.

If you meant we already know it's deadline and we already know the priority so we can fill those directly without doing the extra syscall then yeah. But this is more robust to avoid accidentally changing an attribute that we didn't intend to modify?

Reply:
> The values in sched_data_t must be the ones set to the task or there is a bug. So it's not more robust to call sched_getattr but it's more time consuming

This doesn't make sense to me. sched_getattr() is the robust way to fill the struct sched_attr. Doing it manually encourages more programming mistakes since what we really care about here is maintain the status queue; bar modify the uclamp values. This is the correct way to do RMW as far as I can tell. Relying on the cached values is error prone since it's not  a single valued structure and could grow with time. So we are adding more complexity and increased probability of breakage - for no real gain. The syscall should be pretty fast and the uclamp values shouldn't be changed a lot except for tests that want to stress uclamp code paths.

Reply:
the uclamp fields have been added in sched_data_t, set_thread_priority and parse_sched_data so they have to reuse the fields which have been already filled with relevant informations.
Then syscall is always more costly than reading the value from a field

Reply:
I don't understand the basis of your objection still. How is this breaking anything? I don't see the syscall a performance bottleneck; and the logic is correct as it's doing the RMW in a robust and future proof way.

If the author chose to do it the way you suggest then I'm fine with this. But I don't see a strong technical merit to one way over the other and I think we should respect the author's preference.

Reply:
I have the following change; does this help getting over this deadlock?

```
diff --git a/src/rt-app.c b/src/rt-app.c
index 040cadf..1780518 100644
--- a/src/rt-app.c
+++ b/src/rt-app.c
@@ -982,23 +982,6 @@ static void set_thread_priority(thread_data_t *data, sched_data_t *sched_data)
                sa_params.sched_flags = SCHED_FLAG_KEEP_ALL;
                tid = gettid();

-               /*
-                * Currently we cannot use the syscall to set a clamp
-                * value without specifing a valid policy and priotiry.
-                * Read the current values and use them for the following
-                * syscall.
-                * TODO: fix __sched_setscheduler() to get rid of this!
-                */
-               ret = sched_getattr(tid, &sa_params,
-                                   sizeof(struct sched_attr), flags);
-               if (ret != 0) {
-                       log_critical("[%d] sched_getattr returned %d",
-                                       data->ind, ret);
-                       errno = ret;
-                       perror("sched_getattr");
-                       exit(EXIT_FAILURE);
-               }
-
                if (sched_data->util_min != -1) {
                        sa_params.sched_util_min = sched_data->util_min;
                        sa_params.sched_flags |= SCHED_FLAG_UTIL_CLAMP_MIN;
@@ -1018,6 +1001,13 @@ static void set_thread_priority(thread_data_t *data, sched_data_t *sched_data)
                                   sched_data->util_max);
                }

+               /*
+                * we can't change clamp values without valid policy and
+                * prioroity values.
+                */
+               sa_params.sched_policy = sched_data->policy;
+               sa_params.sched_priority = sched_data->prio;
+
                ret = sched_setattr(tid, &sa_params, flags);
                if (ret != 0) {
                        log_critical("[%d] sched_setattr returned %d",
```

Reply:
I still have to validate it all works as expected. I prefer to push this as a separate PR on top of this though. I am carrying other changes to remove the --with-deadline option from configure and always build with deadline by default.

Reply:
#91 contains all the changes I mentioned. Hopefully this will help to get this merged ASAP. The only issue I can see with this PR as that it can fail to build if building without deadline for the reason mentioned by @credp; if you can edit the file before merging to remove the DLSCHED guards around declaration of sa_params in set_scheduler_priority() that will be great. My PR addresses the issue anyway by removing the optional build of deadline.